operator/src/lib.rs

```
@@ -0,0 +1,269 @@
+}
+
+async fn update_deployment(
+    zk_cluster: &ZooKeeperCluster,
+    context: &Context<ContextData>,
+) -> Result<ReconcilerAction, Error> {
+    let zk_spec: ZooKeeperClusterSpec = zk_cluster.spec.clone();
+    let client = context.get_ref().client.clone();
+
+    let name = Meta::name(zk_cluster);
+    let ns = Meta::namespace(zk_cluster).expect("ZooKeeperCluster is namespaced");
+
+    let mut labels = BTreeMap::new();
+    labels.insert("zookeeper-name".to_string(), name.clone());
+
+    let mut options = HashMap::new();
+    options.insert("tickTime".to_string(), "2000".to_string());
+    options.insert("dataDir".to_string(), "/tmp/zookeeper".to_string()); // TODO: Agent needs to know that this must exist (?) and belong to proper user
+    options.insert("initLimit".to_string(), "5".to_string());
+    options.insert("syncLimit".to_string(), "2".to_string());
+    options.insert("clientPort".to_string(), "2181".to_string());
+
+    for (i, server) in zk_spec.servers.iter().enumerate() {
+        options.insert(
+            format!("server.{}", i + 1),
+            format!("{}:2888:3888", server.node_name),
+        );
+    }
+
+    let mut handlebars = Handlebars::new();
+    handlebars
+        .register_template_string("conf", "{{#each options}}{{@key}}={{this}}\n{{/each}}")
+        .expect("template should work");
+
+    let pods_api: Api<Pod> = Api::namespaced(client.clone(), &ns);
+    let cm_api: Api<ConfigMap> = Api::namespaced(client.clone(), &ns);
+
+    for (i, server) in zk_spec.servers.iter().enumerate() {
+        let pod_name = format!("{}-{}", name, server.node_name);
+        let cm_name = format!("zk-{}", pod_name);
+
+        // This builds the Pod object
+        let pod = build_pod(zk_cluster, server, &labels, &pod_name, &cm_name)?;
+        patch_resource(&pods_api, &pod_name, &pod, FIELD_MANAGER).await?;
+
+        let config = handlebars
+            .render("conf", &json!({ "options": options }))
```

Just a suggestion: I've used strict mode in the agent to ensure we fail early, in case a necessary variable is undefined